src/main/java/edu/stanford/protege/webprotege/issues/EntityDiscussionThreadRepository.java

```diff
@@ -51,27 +56,53 @@ public int getCommentsCount(@Nonnull ProjectId projectId,
 
     public int getOpenCommentsCount(@Nonnull ProjectId projectId,
                                     @Nonnull OWLEntity entity) {
+        List<Comment> comments = getOpenedCommentsForProject(projectId).get(entity.getIRI());
+        return comments == null ? 0 : comments.size();
+    }
+
+
+    public synchronized Map<IRI, List<Comment>> getOpenedCommentsForProject(ProjectId projectId) {
+        Map<IRI, List<Comment>> response = this.cache.get(projectId);
+        if(response != null) {
+            return response;
+        } else {
+            response = new HashMap<>();
+        }
+
         var query = query(where(PROJECT_ID).is(projectId)
-                      .and(ENTITY).is(entity).and(STATUS).is(Status.OPEN));
-        return mongoTemplate.find(query, EntityDiscussionThread.class)
-                        .stream().map(thread -> thread.getComments().size())
-                        .reduce((left, right) -> left + right)
-                        .orElse(0);
+                .and(STATUS).is(Status.OPEN));
+        Map<IRI, List<Comment>> finalResponse = response;
+        mongoTemplate.find(query, EntityDiscussionThread.class)
+                .forEach(entityDiscussionThread -> {
+                    List<Comment> existingComments = finalResponse.get(entityDiscussionThread.getEntity().getIRI());
+                    if(existingComments == null) {
+                        existingComments = new ArrayList<>();
+                    }
+                    existingComments.addAll(entityDiscussionThread.getComments());
+                    finalResponse.put(entityDiscussionThread.getEntity().getIRI(), existingComments);
+                });
+        this.cache.put(projectId, response);
+        return response;
     }
 
+
     public void saveThread(@Nonnull EntityDiscussionThread thread) {
         mongoTemplate.save(thread);
     }
 
     public void addCommentToThread(@Nonnull ThreadId threadId,
+                                   @Nonnull ProjectId projectId,
                                    @Nonnull Comment comment) {
         var query = createQueryForThread(threadId);
+        this.cache.remove(projectId);
```

Comment:
cache bust. when changing a comment on prj , then we bust the cache so when we fetch , we fetch them from the db.